File: 2020/11/2020-11.jl

td_string = """
L.LL.LL.LL
LLLLLLL.LL
L.L.L..L..
LLLL.LL.LL
L.LL.LL.LL
L.LLLLL.LL
..L.L.....
LLLLLLLLLL
L.LLLLLL.L
L.LLLLL.LL"""

read_data(inp) = reduce(vcat, permutedims.(collect.(split(strip(inp), "\n"))))

function nadj(m, sx, sy)
    h,w = size(m)
    c = 0
    for dy = -1:1
        y = sy + dy
        if y >= 1 && y <= h
            for dx = -1:1
                if dy == 0 && dx == 0 continue end
                x = sx + dx
                if x >=1 && x <= w
                    if m[y, x] == '#'
                        c += 1
                    end
                end
            end
        end
    end
    c
end

t1 = read_data("""
               .#.
               ###
               #.#""")
@assert nadj(t1, 2, 2) == 5
@assert nadj(t1, 1, 1) == 3

function life_step(m, seatcount, mxocc)
    h,w = size(m)
    nc = 0
    m2 = copy(m)
    for y = 1:h
        for x = 1:w
            c = m[y, x]
            if c == '.' continue end
            occ = seatcount(m, x, y)
            if c == 'L' && occ == 0
                m2[y, x] = '#'
                nc += 1
            elseif c == '#' && occ >= mxocc 
                m2[y, x] = 'L'
                nc += 1
            end
        end
    end
    m2, nc
end

function part1(m)
    nc = 1
    while nc != 0
        m,nc = life_step(m, nadj, 4)
    end
    count(c->c=='#', m)
end

function firstseat(m, sx, sy)
    h,w = size(m)
    nc = 0
    for dy = -1:1
        for dx = -1:1
            if dy == 0 && dx == 0 continue end
            x,y = (sx + dx, sy + dy)
            while x >= 1 && x <= w && y >= 1 && y <= h
                c = m[y, x]
                if c == '#' || c == 'L'
                    nc += if c == '#' 1 else 0 end
                    break
                end
                x,y = (x + dx, y + dy)
            end
        end
    end
    nc
end

t2 = """
.##.##.
#.#.#.#
##...##
...L...
##...##
#.#.#.#
.##.##."""

td2 = read_data(t2)
@assert firstseat(td2, 4, 4) == 0

function part2(m)
    nc = 1
    while nc != 0
        m,nc = life_step(m, firstseat, 5)
    end
    count(c->c=='#', m)
end

td = read_data(td_string)
@assert part1(td) == 37
data = read_data(read("input.txt", String))
println("PART1: ", part1(data))

@assert part2(td) == 26
println("PART2: ", part2(data))

File: 2020/11/2020-11.exs
ExUnit.start
defmodule Advent11.Test do
  use ExUnit.Case
  def test_data do
    ~S"""
    L.LL.LL.LL
    LLLLLLL.LL
    L.L.L..L..
    LLLL.LL.LL
    L.LL.LL.LL
    L.LLLLL.LL
    ..L.L.....
    LLLLLLLLLL
    L.LLLLLL.L
    L.LLLLL.LL
    """
  end

  def read_data(inp) do
    for {line,r} <- String.split(inp) |> Enum.with_index(),
      {ch, c} <- String.to_charlist(line) |> Enum.with_index(),
      into: %{}, do: {{r,c}, ch}
  end

  def map_to_string(m) do
    for r <- 0..9 do
      (for c <- 0..9, into: [], do: m[{r,c}]) |> to_string()
    end |> Enum.join("\n")
  end

  def num_neighbors(m, r, c) do
    (for dr <- -1..1,
      dc <- -1..1,
      !(dr == 0 && dc == 0), do:
      Map.get(m, {r+dr,c+dc}, ?L)) |> Enum.count(&(&1 == ?#))
  end

  def one_gen({_, m}) do
    m2 = for {{r,c},v} <- m, into: %{} do
      nn = num_neighbors(m, r, c)
      v2 = cond do
        (v == ?L and nn == 0) -> ?#
        (v == ?# and nn >= 4) -> ?L
        true -> v
      end
      {{r,c},v2}
    end
    nc = Enum.count(m2, fn {k,v} -> v != m[k] end)
    {nc, m2}
  end

  def part1(data) do
    {_nc, m} = {1, data}
                |> Stream.iterate(fn {nc,m} -> one_gen({nc,m}) end)
                |> Enum.find(fn {nc,_nf} -> nc == 0 end)
    Map.values(m) |> Enum.count(fn c -> c == ?# end)
  end

  def part2(_data) do
  end

  test "test" do
    td = read_data(test_data())
    assert part1(td) == 37
    data = read_data(File.read!("input.txt"))
    IO.puts "PART1: #{part1(data)}"

    #assert part2(td) == 8
    #assert part2(td2) == 19208
    #IO.puts "PART2: #{part2(data)}"
  end
end
